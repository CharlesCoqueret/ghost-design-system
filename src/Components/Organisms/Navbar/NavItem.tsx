import React, { ReactElement, useRef, useState } from 'react';
import { ControlledMenu, MenuItem } from '@szhsin/react-menu';

import { Icon, IconProp } from '../../Atoms/Icon';
import { Badge, BadgeColorsEnum } from '../../Atoms/Badge';
import { Portal } from '../../Atoms/Portal';

export interface INavItemProps {
  /** Counter in a badge floating above the NavItem (optional, default: undefined) */
  counter?: string | number;
  /** custom submenu (optional, default: undefined) */
  customSubItem?: ReactElement;
  /** Icon on the left of the clickable button (optional if label is defined, default: undefined) */
  icon?: IconProp;
  /** Label of the menu (optional if icon is defuned, default: undefined) */
  label?: string;
  /** Link for redirection on click (optional, default: undefined) */
  link?: string;
  /** On Click handler (optional, default: undefined) */
  onClick?: () => void;
  /** Submenu items (optional, default: undefined) */
  subItems?: Array<INavItemProps>;
}

const NavItem = (props: INavItemProps): ReactElement => {
  const { counter, customSubItem, icon, label, onClick, subItems } = props;

  const ref = useRef<HTMLDivElement>(null);
  const [isOpen, setIsOpen] = useState<boolean>(false);
  const skipOpen = useRef(false);
  const hasMenu = (subItems && subItems.length > 0) || customSubItem !== undefined;

  const closeMenu = () => {
    setIsOpen(false);
  };

  const toggleMenu = () => {
    if (hasMenu) {
      if (!skipOpen.current) {
        setIsOpen((prev) => !prev);
      }
    }
  };

  const handleClick = () => {
    toggleMenu();
    if (onClick) {
      onClick();
    }
  };

  return (
    <>
      <div className='nav-bar-menu-item' ref={ref} onClick={handleClick}>
        {icon && <Icon icon={icon} size='2x' />}
        {label && <div className='nav-bar-menu-label'>{label}</div>}
        {counter && (
          <Badge color={BadgeColorsEnum.DANGER} type='notification' className='counter'>
            {counter}
          </Badge>
        )}
        {hasMenu && label && <Icon icon={['fas', 'caret-down']} size='lg' className='nav-bar-menu-caret' />}
      </div>
      {hasMenu ? (
        <Portal>
          <ControlledMenu
            state={isOpen ? 'open' : 'closed'}
            align='center'
            arrow
            anchorRef={ref}
            skipOpen={skipOpen}
            onClose={closeMenu}>
            {subItems?.map((item): ReactElement => {
              return (
                <MenuItem
                  key={item.label}
                  onClick={() => {
                    if (item.onClick) {
                      item.onClick();
                    }
                  }}>
                  {item.label}
                </MenuItem>
              );
            })}
            {customSubItem}
          </ControlledMenu>
        </Portal>
      ) : (
        <></>
      )}
    </>
  );
};

export default NavItem;
