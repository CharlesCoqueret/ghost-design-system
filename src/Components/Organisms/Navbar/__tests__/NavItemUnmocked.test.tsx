import React from 'react';
import { act, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import NavItem from '../NavItem';

jest.unmock('@szhsin/react-menu');
jest.unmock('react-dom');
describe('NavItem unmocked Component', () => {
  it('NavItem unmocked submenu gets closed', () => {
    const onClickMock = jest.fn();

    const { baseElement } = render(
      <NavItem
        label='ITEM'
        dataTestId='DATA-TEST-ID'
        subItems={[
          {
            counter: 1,
            label: 'SUBITEM',
            icon: ['fal', 'icons'],
            onClick: onClickMock,
            dataTestId: 'SUBITEMDATA-TEST-ID',
          },
        ]}
      />,
    );

    expect(baseElement).toMatchSnapshot();
    expect(onClickMock).not.toBeCalled();

    const item = screen.getByTestId('DATA-TEST-ID');

    act(() => {
      userEvent.click(item);
    });

    expect(baseElement).toMatchSnapshot();
    expect(onClickMock).not.toBeCalled();

    const subitem = screen.getByTestId('SUBITEMDATA-TEST-ID');

    act(() => {
      userEvent.click(subitem);
    });

    expect(baseElement).toMatchSnapshot();
    expect(onClickMock).toBeCalledTimes(1);

    act(() => {
      userEvent.click(item);
    });

    expect(baseElement).toMatchSnapshot();
    expect(onClickMock).toBeCalledTimes(1);
  });
});
