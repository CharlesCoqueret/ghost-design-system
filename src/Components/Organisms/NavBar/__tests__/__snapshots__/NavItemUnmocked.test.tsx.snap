// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavItem unmocked Component NavItem unmocked submenu gets closed 1`] = `
<body>
  <div>
    <div
      class="item"
      data-testid="DATA-TEST-ID"
    >
      <div
        class="label"
      >
        ITEM
      </div>
      <i
        class="icon caret"
        icon="fas,caret-down"
      />
    </div>
  </div>
  <div
    id="root-portal-id"
  >
    <div>
      <div
        class="szh-menu-container"
        style="position: relative;"
      >
        <ul
          aria-label="Menu"
          class="szh-menu szh-menu--state-closed szh-menu--dir-bottom"
          role="menu"
          style="position: absolute; margin: 0px; display: none; left: 0px; top: 0px;"
          tabindex="-1"
        >
          <div
            style="position: absolute; left: 0px; top: 0px;"
            tabindex="-1"
          />
          <div
            class="szh-menu__arrow szh-menu__arrow--dir-bottom"
            style="position: absolute;"
          />
          <a
            href="#"
          >
            partial url
            <li
              class="szh-menu__item"
              data-testid="SUBITEMDATA-TEST-ID"
              role="menuitem"
              tabindex="-1"
            >
              SUBITEM
            </li>
          </a>
        </ul>
      </div>
    </div>
  </div>
</body>
`;

exports[`NavItem unmocked Component NavItem unmocked submenu gets closed 2`] = `
<body>
  <div>
    <div
      class="item"
      data-testid="DATA-TEST-ID"
    >
      <div
        class="label"
      >
        ITEM
      </div>
      <i
        class="icon caret"
        icon="fas,caret-down"
      />
    </div>
  </div>
  <div
    id="root-portal-id"
  >
    <div>
      <div
        class="szh-menu-container"
        style="position: relative;"
      >
        <ul
          aria-label="Menu"
          class="szh-menu szh-menu--state-open szh-menu--dir-bottom"
          role="menu"
          style="position: absolute; margin: 0px; left: 0px; top: 0px;"
          tabindex="-1"
        >
          <div
            style="position: absolute; left: 0px; top: 0px;"
            tabindex="-1"
          />
          <div
            class="szh-menu__arrow szh-menu__arrow--dir-bottom"
            style="position: absolute; left: 0px;"
          />
          <a
            href="#"
          >
            partial url
            <li
              class="szh-menu__item"
              data-testid="SUBITEMDATA-TEST-ID"
              role="menuitem"
              tabindex="-1"
            >
              SUBITEM
            </li>
          </a>
        </ul>
      </div>
    </div>
  </div>
</body>
`;

exports[`NavItem unmocked Component NavItem unmocked submenu gets closed 3`] = `
<body>
  <div>
    <div
      class="item"
      data-testid="DATA-TEST-ID"
    >
      <div
        class="label"
      >
        ITEM
      </div>
      <i
        class="icon caret"
        icon="fas,caret-down"
      />
    </div>
  </div>
  <div
    id="root-portal-id"
  >
    <div>
      <div
        class="szh-menu-container"
        style="position: relative;"
      >
        <ul
          aria-label="Menu"
          class="szh-menu szh-menu--state-closed szh-menu--dir-bottom"
          role="menu"
          style="position: absolute; margin: 0px; left: 0px; top: 0px; display: none;"
          tabindex="-1"
        >
          <div
            style="position: absolute; left: 0px; top: 0px;"
            tabindex="-1"
          />
          <div
            class="szh-menu__arrow szh-menu__arrow--dir-bottom"
            style="position: absolute; left: 0px;"
          />
          <a
            href="#"
          >
            partial url
            <li
              class="szh-menu__item"
              data-testid="SUBITEMDATA-TEST-ID"
              role="menuitem"
              tabindex="-1"
            >
              SUBITEM
            </li>
          </a>
        </ul>
      </div>
    </div>
  </div>
</body>
`;

exports[`NavItem unmocked Component NavItem unmocked submenu gets closed 4`] = `
<body>
  <div>
    <div
      class="item"
      data-testid="DATA-TEST-ID"
    >
      <div
        class="label"
      >
        ITEM
      </div>
      <i
        class="icon caret"
        icon="fas,caret-down"
      />
    </div>
  </div>
  <div
    id="root-portal-id"
  >
    <div>
      <div
        class="szh-menu-container"
        style="position: relative;"
      >
        <ul
          aria-label="Menu"
          class="szh-menu szh-menu--state-open szh-menu--dir-bottom"
          role="menu"
          style="position: absolute; margin: 0px; left: 0px; top: 0px;"
          tabindex="-1"
        >
          <div
            style="position: absolute; left: 0px; top: 0px;"
            tabindex="-1"
          />
          <div
            class="szh-menu__arrow szh-menu__arrow--dir-bottom"
            style="position: absolute; left: 0px;"
          />
          <a
            href="#"
          >
            partial url
            <li
              class="szh-menu__item"
              data-testid="SUBITEMDATA-TEST-ID"
              role="menuitem"
              tabindex="-1"
            >
              SUBITEM
            </li>
          </a>
        </ul>
      </div>
    </div>
  </div>
</body>
`;
